migrate/live-migration/live-migration-from-timescaledb.md

```diff
@@ -0,0 +1,153 @@
+or use terminal multiplexers like `screen` or `tmux`.
+
+The command will take a snapshot of your source database, migrate existing data to the
+target, and stream live transactions from the source to the target. During this process,
+it will display the lag between the source and target databases in terms of WAL offset size.
+
+```sh
+[WATCH] Source DB - Target DB => 126MB
+```
+
+When the lag between the source and target database is less than 30 megabytes, it will
+start `ANALYZE` on the target database. This updates statistics in the target database,
+which is necessary for optimal querying performance in the target database. Wait for
+`ANALYZE` to complete.
+
+<Highlight type="note">
+Application downtime begins here.
+</Highlight>
+
+When you're _ready to take downtime_ (i.e., stop normal operations) on your production
+applications that are dependent on your source database and the _lag between the databases is below 30 megabytes_,
+you should stop your applications that perform DML queries on your source database.
+This is the downtime phase and will last until you have completed the validation step (4).
+Be sure to go through the validation step (4) and prepare the validation queries before
+you enter the downtime phase to keep the overall downtime minimal.
+
+Stopping DML queries on your source database allows the live migration image to make
+the target database even with the source. This will be evident when the lag outputs
+0 megabytes.
+
+When the lag between the two databases shows 0 megabytes, wait for a few minutes and
+then provide the signal to stop the live replay by pressing key `s` or giving a
+`SIGUSR1` signal, as per your log. You should see `Stopped` message.
+
+```sh
+[WATCH] Source DB - Target DB => 0MB. Press 's' (and ENTER) to stop live-replay
+Stopped
+```
+
+Now, compare tables in the target against those in the source to ensure that
+both have the same data.
+
+
+Once you are confident that the tables in the source are identical to the target,
+you need to signal "continue". This is done either by pressing key `c` or giving a `SIGUSR1`
+signal, as mentioned in your log.
+
+```sh
+[ACTION NEEDED] Now, you should check the integrity of your data. Once you are confident, press 'c' (and ENTER) to continue
+Continuing ...
+Syncing last LSN in Source DB to Target DB ...
+```
+
+The live migration image will continue the remaining work under live replay,
+copy TimescaleDB metadata, sequences, and run policies. You should see the
+following message if all the mentioned steps were successful.
+
+```sh
+Migration successfully completed
+```
+
+## 4. Validate the data in target database and promote as new primary
+
+Now that all data has been migrated, the contents of both databases should be the
+same. How exactly this should best be validated is dependent on your application.
+You could compare the number of rows or an aggregate of columns to validate that
+the target database matches with the source.
```

Comment:
This 2 sections talk about validation. I still think that the wait for confirmation to check the integrity is not necessary, specially since we are asking them to validate again when the migration ends.

My gripe with this, is that we leave it for the customer to find out what validation is, which might add more stress to the process (specially since it's on the downtime phase).

What about adding telling them to validate while the data is being replicated, like run some queries with an upper bound that's shared between target and source while the migration is going. You can re-run them again at the end, using the rest of the time.

Reply:
@jgpruitt what did you use to do for validation when doing replication? Did you just assume all the data was there? run some queries?

Reply:
> My gripe is that we leave it for the customer to find out what validation is, which might add more stress to the process (specially since it's on the downtime phase).

I understand the point and agree to it. I want to add that we probably need a tool for verifying data across different databases, because users using pg_dump & pg_restore, and dual-writes backfill will also face the same situation.

> What about adding telling them to validate while the data is being replicated, like run some queries with an upper bound that's shared between target and source while the migration is going.

This may not be right, because there are chances for DML in compressed chunks/old data that will show up as momentary inconsistencies (until it's been streamed to target).

Reply:
> I still think that the wait for confirmation to check the integrity is not necessary, specially since we are asking them to validate again when the migration ends.

Hmm, giving this another thought — we basically ask for 2 signals post starting live migration. They are `s` and `c` key press events.

`s` is asked because
1. We want to know at what moment is the user ready to perform the final step of migration (the step that migrates caggs from source to target). This is important as after this signal, downtime starts.
2. We give user a chance to check if initial data + live-replay data migration matches with source or not. This is mostly for safety reasons based on what I recall.

`c` is asked mainly to continue the remaining steps in migration process. These include copying sequences, caggs, and related work.

Now, the time lapse between receiving `c` and completing the migration is very small.

So, I do agree we can remove the part of validation between `s` and `c` since the time between `c` to end is relatively negligible, so if the data is not consistent, users will know it after the end of migration. Asking them to do the validation few minutes earlier (between `s` & `c`) does not make much sense.


How about this — after live migration starts, we just accept `c` only. Users will provide `c` key press when the _delay between source and target is 0_ and the _user is ready to take downtime_, meaning, we can stop live-replay and proceed with copying sequences, caggs, and remaining work.

@alejandrodnm @arajkumar WDYT?

Reply:
@arajkumar and I discussed about this, and we agreed that we should allow users to check data consistency at the end of migration process. ATM, user have to validate data twice — first, when he gives `s`, he checks for existing data + live replay, but without caggs (since they are yet to be migrated); second time, after migration completes, he verifies for the entire database. With the new approach, user will only perform verification on entire database after the migration completes. Hence, we will keep `c`, and remove the `s` signal.

For instance, the new workflow will be as follows.

User starts live migration for TS to TS. When the delay between source and target is 0 (he may have to stop load on the source side) and the user is ready to take the "minimal downtime", he will press `c` key. The script will then stop live-replay and do the remaining work to complete the migration process. 

If the migration is for PG to TS, an additional step for `create_hypertable()` will be asked before the `c` key press.

Reply:
Update: With https://github.com/timescale/pgcopydb/pull/22 merged, the live migration image behaviour aligns with the above mentioned approach.